bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BStreamingJSON.java

```diff
@@ -68,11 +80,18 @@ public void serialize(OutputStream outputStream) {
          */
         try {
             JsonGenerator gen = new JsonGenerator(outputStream);
-            if (this.values != null) {
-                gen.serialize(this);
-            } else {
-                this.datasource.serialize(gen);
+            gen.writeStartArray();
+
+            // First serialize the values loaded to memory
+            for (int i = 0; i < size; i++) {
+                gen.serialize(values[i]);
+            }
```

Comment:
Could have used a foreach loop here.

Reply:
We actually cannot. `values` array is an auto growing array, and is always larger than the actual filled-size of the array. Hence using foreach will print the un-filled part of the array as well.